DEVX-1580 Provide some initial toolchain enhancements

bbejeck opened the pull request:
This PR provides an initial attempt to provide some utilities for tutorial authors.

Here's the TLDR;

`gen_project.sh` 

Generates the project scaffold with as much content as feasible.  To use
```
./gen_project.sh /some/directory/my_tutorial_properties.sh
```
Where `my_tutorial_properties.sh` would look like
```
#!/bin/sh

# space separated list of types, could just be one of ksql or kstreams
CARDS="kstreams ksql"

# all lower-case with no spaces
TUTORIAL_SHORT_NAME=my-tutorial-name

# also will become a test named FilteringCodeTest
MAIN_CLASS=FilteringCode
AK_VERSION=2.4.0
CP_VERSION=5.4.0
KSQLDB_VERSION=0.7.1
SEMAPHORE_TEST_NAME="Test name for Semaphore run"
PERMALINK="seo-friendly-link-to-my-tutorial"
```

`clone-tutorial.sh`
Clones an existing tutorial, changing the tutorial name in the content.  Great for new tutorial that similar to exiting ones.

```
./clone-tutorial.sh some/directory/clone_propteries.sh
```
Sample clone properties file
```
#!/bin/sh

# clones everything ksql and kstreams
ORIG_TUTORIAL=filtering

# to clone just the ksql part
#ORIG_TUTORIAL=filtering/ksql

# to clone just the kstreams portion
#ORIG_TUTORIAL=filtering/kstreams
NEW_TUTORIAL=something-close-to-filtering
SEMAPHORE_TEST_NAME="My New Filtering Tutorial"
PERMALINK=filtering-new-hotness

colinhicks commented on templates/ksql/filtered/code/docker-compose.yml:
```diff
@@ -0,0 +1,80 @@
+
+  ksql-server:
+    image: confluentinc/cp-ksql-server:<CP-VERSION>
+    hostname: ksql-server
```
I presume the plan would be to tweak this for ksqlDB and ask for a KSQLDB_VERSION?

bbejeck replied:
Yes, I missed that one.

bbejeck commented:
@colinhicks updated and rebased on master

colinhicks commented:
Thanks @bbejeck, this looks fantastic. Looking forward to kicking the tires.

bbejeck commented:
When building a tutorial with `./gen_project.sh` CD into `_includes/<TUTORIAL-NAME>/kstreams/code`  then run

1. `gradle wrapper`
2. `./gradlew build`
3. `/.gradlew idea`
 Then open the project in IntelliJ.  More to come..

colinhicks commented on templates/ksql/filtered/code/docker-compose.yml:
```diff
@@ -0,0 +1,80 @@
+      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://broker:9092,PLAINTEXT_HOST://localhost:29092
+      KAFKA_METRIC_REPORTERS: io.confluent.metrics.reporter.ConfluentMetricsReporter
+      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
+      KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS: 0
```
```suggestion
      KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS: 0
      KAFKA_TRANSACTION_STATE_LOG_MIN_ISR: 1
      KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 1
```

colinhicks commented on templates/ksql/static/dev/code/start-cli.sh:
```diff
@@ -0,0 +1 @@
+docker exec -it ksql-cli ksql http://ksql-server:8088
```
```suggestion
docker exec -it ksqldb-cli ksql http://ksql-server:8088
```

bbejeck commented:
Now running `gen_project.sh` with `kstreams` produces a fully running bare-bones tutorial, next is `ksql`

bbejeck commented:
Updated with:
1. Running `ksql` template project with generation
2. Prints a checklist at the end of the generation or cloning process plus prints a copy of the checklist in the new tutorial directory
3. Moved generation and cloning scripts to a `tools` directory

bbejeck commented:
@colinhicks after you review the `README` I think we have something we can merge.

colinhicks commented on README.md:
```diff
@@ -78,6 +78,111 @@ Kafka Tutorials is a bit unique in that each tutorial is self-testing. That is,
+
+As the name implies, the `gen_project.sh` script generates the minimal structure for a viable Kafka Tutorials tutorial.  The code and text contained in the tutorial are place holders, and you'll update those with your code and writing to complete the tutorial.
+
+The script expects you to supply a properties file for setting the name and other parts of the tutorial.  Heres' an example properties file you can use.
```
```suggestion
The script expects you to supply a properties file for setting the name and other parts of the tutorial.  Here's an example properties file you can use.
```

colinhicks commented on README.md:
```diff
@@ -78,6 +78,111 @@ Kafka Tutorials is a bit unique in that each tutorial is self-testing. That is,
+
+1. Create a branch
+2. Make sure you are in the `Kafka Tutorials` base directory
+3. Execute `> tools/clone_tutoral.sh ~/clone-tutorial-props.sh`
```
```suggestion
3. Execute `> tools/clone_tutorial.sh ~/clone-tutorial-props.sh`
```
